test/framework-mocha.js

```diff
@@ -0,0 +1,420 @@
+                                  'vnis', 'deleteAllVnis', 'deleteVni', 'vni' );
+
+        // Verify we got the ports we expect to have
+        Object.keys( node.inPorts ).should.have.length(1);
+        node.inPorts.should.have.all.keys( 'input1' );
+        Object.keys( node.outPorts ).should.have.length(2);
+        node.outPorts.should.have.all.keys( 'output', 'error' );
+    });
+
+    it( "should support creation of node instances with custom output ports", function() {
+
+        var node = test.createComponent( framework.componentFactory({
+            description: "Test Description",
+            inPorts: {
+                input1: {
+                    datatype: 'string',
+                    description: "a string port",
+                    required: true
+                },
+                input2: {
+                    datatype: 'object',
+                    description: "an object port"
+                }
+            },
+            outPorts: {
+                // Define a custom output port
+                my_output: {
+                    datatype: 'string',
+                    description: "a string port",
+                }
+            }
+        }));
+
+        // Verify we got a basic node 
+        node.should.exist;
+        node.should.be.an('object');
+        node.should.include.keys('nodeName', 'componentName', 'inPorts', 'outPorts', 
+                                  'vnis', 'deleteAllVnis', 'deleteVni', 'vni' );
+
+        // Verify we got the ports we expect to have
+        Object.keys( node.inPorts ).should.have.length(2);
+        node.inPorts.should.have.all.keys( 'input1', 'input2' );
+        Object.keys( node.outPorts ).should.have.length(3);
+        node.outPorts.should.have.all.keys( 'my_output', 'output', 'error' );
+    });
+
+    describe("#ondata", function() {
+
+        it( "should manage single port node vni state, fRunUpdater invocation, & output state", function() {
+
+            var inData = "A bit of input data";
+            var executedFRunUpdater = "Executed fRunUpdater API";
+
+            // Define the fRunUpdater that framework should invoke
+            var fRunUpdater = function( inputStates ) { 
+                return executedFRunUpdater;
+            }
+
+            // Create a noflo component and get the node instance for it
+            var node = test.createComponent( framework.componentFactory({
+                description: "Test Description",
+                inPorts: {
+                    input1: {
+                        datatype: 'string',
+                        description: "a string port",
+                        required: true
+                    }
+                }
+            }, fRunUpdater));
+
+            // Send data to the input port and verify that the fRunUpdater function is called.
+            // We use a promise here because this section is asynchronous
+            return new Promise( function(done, fail) { 
```

Comment:
Suggest renaming these parameters to be resolve and reject, in accordance with normal Promises convention, and to distinguish from the 'done' and 'fail' parameters below, which are (I believe) completely different.